File: Client/Roll10/Services/DiceLogService.cs
using System.Reactive.Disposables;
using System.Reactive.Subjects;
using Roll10.Domain.Models;

namespace Roll10.Services
{


    public class DiceLogService: IDisposable
    {
        private readonly PocketbaseService _pb;
        private readonly ToastService _toastService;
        private readonly string _logCollectionName = "diceroomlogs";
        public List<DiceLogEntry> DiceLog { get; set; } = new();
        private bool _hasSynced;

        public Subject<(DiceLogEntry, bool)> AddDiceLogEntrySubject { get; } = new();
        private readonly List<IDisposable> _subs;

        public DiceLogService(PocketbaseService pb, ToastService toastService)
        {
            _pb = pb;
            _toastService = toastService;
            _subs = new List<IDisposable>
            {
                AddDiceLogEntrySubject.Subscribe(values =>
                {
                    _ = Task.Run(async () =>
                    {
                        await AddEntry(values.Item1, values.Item2);
                    });
                })
            };
        }

        public async Task SyncDiceLog()
        {
            if (!_hasSynced)
            {
                var events = await _pb.GetFullList<DiceLogEntry>(_logCollectionName);
                var currentIds = DiceLog.Select(d => d.id);
                events = events.Where(e => !currentIds.Contains(e.id)).ToList();
                DiceLog.AddRange(events);
                _hasSynced = true;
            }
        }

        public void ClearLog()
        {
            _hasSynced = false;
            DiceLog = new();
        }

        private async Task AddEntry(DiceLogEntry entry, bool directPush = false)
        {
            var user = await _pb.GetUser();
            if(!string.IsNullOrEmpty(user?.diceroom) && !directPush)
            {
                await _pb.CreateItem(_logCollectionName, entry with { room_id = user.diceroom });
                return;
            }

            if (!DiceLog.Select(d => d.id).Contains(entry.id))
            {
                DiceLog.Add(entry);
                await _toastService.ShowToast(new ToastMessage(
                    ToastType.Info,
                    $"<p class=\"text-center font-bold\">{entry.title}</p><p class=\"text-center italic\">{entry.diceroll}</p><p class=\"text-center font-bold text-lg\">{entry.rolledamount}</p>"
                ));
            }
        }

        public void Dispose()
        {
            AddDiceLogEntrySubject.Dispose();
            foreach (var sub in _subs) sub.Dispose();
        }
    }   
}

File: Client/Roll10Angular/src/app/dice-log.service.ts
import { Injectable } from '@angular/core';
import { Subject, Subscription } from 'rxjs';
import { PocketBaseService } from './pocket-base.service'
import { IDiceLogEntry } from "../domain/data/DiceLogEntry";
import { IToastMessage, ToastService } from './toast.service';
import * as _ from "lodash";
import { GenerateId } from 'src/Helpers';

const collectionName = "diceroomlogs";

export interface IDiceLogPush {
  entry:IDiceLogEntry,
  directPush: boolean
}

@Injectable({
  providedIn: 'root'
})
export class DiceLogService {

  public addDiceLogSubject = new Subject<IDiceLogPush>()
  private hasSynced = false;

  private subs: Subscription[] = [];
  public diceLog: IDiceLogEntry[] = [];

  constructor(
    private pb: PocketBaseService,
    private toast: ToastService
  ) 
  {
    this.subs.push(this.addDiceLogSubject.subscribe(next => {
      this.AddEntry(next);
    }))
  }

  public async SyncDiceLog()
  {
    if(!this.hasSynced)
    {
      let events = await this.pb.GetFullList<IDiceLogEntry>(collectionName);
      let currentIds = this.diceLog.map(d => d.id);
      this.diceLog = _.chain(this.diceLog)
                      .concat(events.filter(e => !currentIds.includes(e.id)))
                      .sortBy("created").value();

      this.hasSynced = true;
    }
  }

  public ClearLog()
  {
    this.hasSynced = false;
    this.diceLog = [];
  }

  private async AddEntry(push: IDiceLogPush)
  {
    let user = this.pb.GetUser();
    if(user?.diceroom && !push.directPush)
    {
      await this.pb.CreateItem(collectionName, {...push.entry, room_id: user.diceroom})
    }

    if(!this.diceLog.map(d => d.id).includes(push.entry.id))
    {
      this.diceLog.push(push.entry);
      let toastMessage:IToastMessage = {
        id: GenerateId(),
        type: 'Info',
        message: `<p class=\"text-center font-bold\">${push.entry.title}</p><p class=\"text-center italic\">${push.entry.diceroll}</p><p class=\"text-center font-bold text-lg\">${push.entry.rolledamount}</p>`
      }
      this.toast.ShowToast(toastMessage);
    }
  }
}
